main.go

```diff
@@ -44,6 +49,8 @@ func main() {
 		emailClient,
 		ipGeolocation,
 		sessionStore,
+		devRepo,
```

Comment:
We should probably inject these repository dependencies into the respective handlers. So for instance wherever we have 

```
svr.RegisterRoute("/Golang-Developers", handler.DevelopersHandler(svr), []string{"GET"})
```
it becomes
```
svr.RegisterRoute("/Golang-Developers", handler.DevelopersHandler(svr, devRepo), []string{"GET"})
```

The final objective is to have at some point handlers that take only the minimum number of necessary dependencies and hopefully get rid of the "god object" server struct which is now being passed everywhere. In this way we avoid coupling too much dependencies and logic into a single object (server.go)

So not a big change in this file, just have to individually pass the repositories to each handler instead of injecting that into the server.

Reply:
This allows us to pass along composable components (repository, email sender, etc) once we start cleaning all up and splitting out logic from the big server.go file

Reply:
What do you think is a better approach? Injecting components like email client, repo, loggers, etc to a Handler struct like:
```go
// developer/handler.go

type logger interface {
	Log(err error, msg string)
}

type renderer interface {
	Render(w http.ResponseWriter, status int, htmlView string, data interface{}) error
}

type emailer interface {
	SendEmail(from, to, replyTo, subject, text string) error
}

type Handler struct {
	repo *Repository
	logger
	renderer
	emailer
}

func NewHandler(repo *Repository, l logger, rend renderer, e emailer) *Handler {
	return &Handler{
		repo:     repo,
		logger:   l,
		renderer: rend,
		emailer:  e,
	}
}

func (h *Handler) SaveDeveloperProfile(userRepo *user.Repository) http.HandlerFunc {
     // ...
}

// main.go

func main() {

	// ...

	devRepo := developer.NewRepository(conn)
	devHandler := developer.NewHandler(devRepo, logger, template.NewTemplate(), emailClient)

       // ...

     svr.RegisterRoute("/x/sdp", devHandler.SaveDeveloperProfile(userRepo), []string{"POST"})
```

Or passing components to a handler func directly like:

```go
// developer/handler.go

type logger interface {
	Log(err error, msg string)
}

type renderer interface {
	Render(w http.ResponseWriter, status int, htmlView string, data interface{}) error
}

type emailer interface {
	SendEmail(from, to, replyTo, subject, text string) error
}

type Handler struct {
	repo *Repository
}

func NewHandler(repo *Repository) *Handler {
	return &Handler{
		repo:     repo,
	}
}

func (h *Handler) SaveDeveloperProfile(userRepo *user.Repository, e emailer, l logger) http.HandlerFunc {
     // ...
}

// main.go

func main() {

	// ...

	devRepo := developer.NewRepository(conn)
	devHandler := developer.NewHandler(devRepo)

       // ...

     svr.RegisterRoute("/x/sdp", devHandler.SaveDeveloperProfile(userRepo, logger, emailClient), []string{"POST"})
```

Reply:
I think for now the easiest option is to not create a handler struct and pass them directly into the method call. But you don't need to de-couple everything right now, just pass the repository and the server. Then bit by bit we can pull the remaining dependencies (email sender, renderer etc) out of the server object. So something like this

in `main.go`
```
devRepo := developer.NewRepository(conn)

// ...

svr.RegisterRoute("/x/sdp", handler.SaveDeveloperProfile(svr, devRepo), []string{"POST"})
```

in `handlers.go`
```
type devGetter interface {
  GetDeveloper(string) Developer
}

func SaveDeveloperProfile(svr *server.Server, devRepo devGetter) http.HandlerFunc {
  // ... same as before
}
```

Does it make sense? It's much less things to worry about right now so we keep the PR not too big. Later on we can always pull out other dependencies from server and inject into the handler function

Reply:
Perfect!